src/unix_term.rs

```diff
@@ -90,76 +126,123 @@ pub fn read_single_key() -> io::Result<Key> {
+                            'D' => Ok(Key::ArrowLeft),
+                            'H' => Ok(Key::Home),
+                            'F' => Ok(Key::End),
+                            'Z' => Ok(Key::BackTab),
+                            _ => {
+                                let c3 = read_single_char(fd)?;
+                                if let Some(c3) = c3 {
+                                    if c3 == '~' {
+                                        match c2 {
+                                            '2' => Ok(Key::Insert),
+                                            '3' => Ok(Key::Del),
+                                            '5' => Ok(Key::PageUp),
+                                            '6' => Ok(Key::PageDown),
+                                            _ => Ok(Key::UnknownEscSeq(vec![c1, c2, c3])),
+                                        }
+                                    } else {
+                                        Ok(Key::UnknownEscSeq(vec![c1, c2, c3]))
+                                    }
+                                } else {
+                                    // \x1b[ and 1 more char
+                                    Ok(Key::UnknownEscSeq(vec![c1, c2]))
+                                }
+                            }
+                        }
+                    } else {
+                        // \x1b[ and no more input
+                        Ok(Key::UnknownEscSeq(vec![c1]))
+                    }
+                } else {
+                    // char after escape is not [
+                    Ok(Key::UnknownEscSeq(vec![c1]))
+                }
             } else {
-                Ok(key_from_escape_codes(&buf[..2 as usize]))
+                //nothing after escape
+                Ok(Key::Escape)
             }
-        } else if buf[0] & 240u8 == 224u8 {
-            // a three byte unicode character
-            let read = libc::read(fd, buf[1..].as_mut_ptr() as *mut libc::c_void, 2);
-            if read < 0 {
-                Err(io::Error::last_os_error())
-            } else if buf[1] == b'\x03' {
-                Err(io::Error::new(
-                    io::ErrorKind::Interrupted,
-                    "read interrupted",
-                ))
+        }
+        Some(c) => {
+            let byte = c as u8;
+            let mut buf: [u8; 4] = [byte, 0, 0, 0];
+
+            if byte & 224u8 == 192u8 {
+                // a two byte unicode character
+                let read = unsafe { libc::read(fd, buf[1..].as_mut_ptr() as *mut libc::c_void, 1) };
+                if read < 0 {
+                    Err(io::Error::last_os_error())
+                } else if buf[1] == b'\x03' {
+                    Err(io::Error::new(
+                        io::ErrorKind::Interrupted,
+                        "read interrupted",
+                    ))
+                } else {
+                    Ok(key_from_escape_codes(&buf[..2 as usize]))
```

Comment:
Do we need to use this function anymore?

Reply:
I removed the character recognition from it. However, I had to rename it to `key_from_utf8` and use it as a helper method when dealing with multibyte chars.